bcl/src/bc_log.c

```diff
@@ -22,76 +28,71 @@ void bc_log_debug(const char *format, ...)
 {
     va_list ap;
 
-    if (_bc_log.level == BC_LOG_LEVEL_OFF || _bc_log.level > BC_LOG_LEVEL_DEBUG)
-    {
-        return;
-    }
-
-    strcpy(_bc_log.buffer, "# <D> ");
-
     va_start(ap, format);
-    vsnprintf(&_bc_log.buffer[6], sizeof(_bc_log.buffer) - 6 - 2, format, ap);
+    _bc_log_message(BC_LOG_LEVEL_DEBUG, 'D', format, ap);
     va_end(ap);
-
-    strcat(_bc_log.buffer, "\r\n");
-
-    bc_uart_write(BC_UART_UART2, _bc_log.buffer, strlen(_bc_log.buffer));
 }
 
 void bc_log_info(const char *format, ...)
 {
     va_list ap;
 
-    if (_bc_log.level == BC_LOG_LEVEL_OFF || _bc_log.level > BC_LOG_LEVEL_INFO)
-    {
-        return;
-    }
-
-    strcpy(_bc_log.buffer, "# <I> ");
-
     va_start(ap, format);
-    vsnprintf(&_bc_log.buffer[6], sizeof(_bc_log.buffer) - 6 - 2, format, ap);
+    _bc_log_message(BC_LOG_LEVEL_INFO, 'I', format, ap);
     va_end(ap);
-
-    strcat(_bc_log.buffer, "\r\n");
-
-    bc_uart_write(BC_UART_UART2, _bc_log.buffer, strlen(_bc_log.buffer));
 }
 
 void bc_log_warning(const char *format, ...)
 {
     va_list ap;
 
-    if (_bc_log.level == BC_LOG_LEVEL_OFF || _bc_log.level > BC_LOG_LEVEL_WARNING)
-    {
-        return;
-    }
-
-    strcpy(_bc_log.buffer, "# <W> ");
-
     va_start(ap, format);
-    vsnprintf(&_bc_log.buffer[6], sizeof(_bc_log.buffer) - 6 - 2, format, ap);
+    _bc_log_message(BC_LOG_LEVEL_WARNING, 'W', format, ap);
     va_end(ap);
-
-    strcat(_bc_log.buffer, "\r\n");
-
-    bc_uart_write(BC_UART_UART2, _bc_log.buffer, strlen(_bc_log.buffer));
 }
 
 void bc_log_error(const char *format, ...)
 {
     va_list ap;
 
-    if (_bc_log.level == BC_LOG_LEVEL_OFF || _bc_log.level > BC_LOG_LEVEL_ERROR)
+    va_start(ap, format);
+    _bc_log_message(BC_LOG_LEVEL_ERROR, 'E', format, ap);
+    va_end(ap);
+}
+
+static void _bc_log_message(bc_log_level_t level, char id, const char *format, va_list ap)
+{
+    if (_bc_log.level == BC_LOG_LEVEL_OFF || _bc_log.level > level)
```

Comment:
What if the debug level was also set by DEFINE during compilation? You can then replace called log_info & log_debug by (void), so the code will be removed. I see that DEBUG/RELEASE does this but I'm afraid of the bloated code with string messages from log_debug, log_info when for example I would like to log only log_error().

Reply:
Ok, we can deal with it for now. We can change it easily in the future if the strings become an issue.